ori/volume/server/volume_create.go

```diff
@@ -17,260 +17,132 @@ package server
-					Kind:     "Volume",
-					Name:     pvc.Name,
-				},
-			},
-		}
-		metautils.SetLabel(imagePvc, volumev1alpha1.ManagerLabel, volumev1alpha1.VolumeCephletManager)
-
+	log.V(2).Info("Resolving os image")
+	img, err := s.populator.ResolveImage(ctx, imageName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve image: %w", err)
 	}
 
-	if err := apiutils.SetObjectMetadata(pvc, volume.Metadata); err != nil {
-		return nil, err
+	if img.RootFS == nil {
+		return nil, fmt.Errorf("failed to get rootFS layer")
 	}
-	metautils.SetLabel(pvc, volumev1alpha1.ManagerLabel, volumev1alpha1.VolumeCephletManager)
-	apiutils.SetVolumeClassLabel(pvc, volume.Spec.Class)
 
-	return &AggregateCephVolume{
-		Pvc:         pvc,
-		ImagePvc:    imagePvc,
-		VolumeClass: volumeClass,
+	return &PopulationImage{
+		Name: imageName,
+		Size: uint64(img.RootFS.Descriptor().Size),
 	}, nil
 }
 
-func (s *Server) handleImagePopulation(ctx context.Context, log logr.Logger, volume *AggregateCephVolume) error {
-	if volume.ImagePvc == nil {
-		return nil
+func (s *Server) validateVolume(ctx context.Context, log logr.Logger, oriVolume *ori.Volume) error {
+	log.V(2).Info("Starting volume validation")
+	defer log.V(2).Info("Finished volume validation")
+
+	if oriVolume == nil {
+		return fmt.Errorf("volume is nil")
 	}
 
-	snapshot := &snapshotv1.VolumeSnapshot{}
-	if err := s.client.Get(ctx, client.ObjectKeyFromObject(volume.ImagePvc), snapshot); err != nil {
-		if !errors.IsNotFound(err) {
-			return fmt.Errorf("unable to get snapshot: %w", err)
-		}
-		log.V(1).Info("Requested snapshot not found: create image pvc and snapshot it.")
-		return s.createSnapshot(ctx, log, volume)
+	if err := s.validateVolumeClass(oriVolume); err != nil {
+		return fmt.Errorf("invalid volume class: %w", err)
 	}
+	log.V(2).Info("Validated class")
 
-	if !pointer.BoolDeref(snapshot.Status.ReadyToUse, false) || snapshot.Status.RestoreSize == nil {
-		log.Info("Referenced snapshot not ready or RestoreSize not defined", "snapshotName", snapshot.Name)
-		return nil
+	validatedImage, err := s.validateImage(ctx, log, oriVolume)
+	if err != nil {
+		return fmt.Errorf("invalid volume image: %w", err)
 	}
+	log.V(2).Info("Validated image")
 
-	volumeSizeBytes := volume.Pvc.Spec.Resources.Requests.Storage().Value()
-	if volumeSizeBytes < snapshot.Status.RestoreSize.Value() {
-		log.Info(fmt.Sprintf("Requested volume size %d is less than the size %d for the source snapshot", volumeSizeBytes, snapshot.Status.RestoreSize.Value()), "snapshotName", snapshot.Name)
-		//r.Eventf(volume, corev1.EventTypeWarning, ReasonVolumeSizeToSmall, "Requested volume size %d is less than the size %d for the source snapshot %s", volumeSizeBytes, snapshot.Status.RestoreSize.Value(), snapshot.Name)
+	if err := s.validateSize(oriVolume, validatedImage); err != nil {
+		return fmt.Errorf("invalid volume size: %w", err)
 	}
+	log.V(2).Info("Validated size")
```

Comment:
Now that I see that the size depends on the image, you cannot do this validation at creation time. You'll have to add this as a runtime error / mark the volume as state == Errored or sth. like that when actually trying to pull.